### ISSUE TITLE
sparse_mask method ignores masked-in elements of sparse compressed input tensors

### ISSUE THREAD

**Issue by pearu** [pytorch/pytorch#107381]
## Issue description

As in the title. The current behavior with sparse compressed inputs contradicts [sparse_mask](https://pytorch.org/docs/main/generated/torch.Tensor.sparse_mask.html#torch.Tensor.sparse_mask) documentation note: "The returned sparse tensor has the same indices as the sparse tensor mask, even when the corresponding values in self are zeros." as well as the current behavior with strided or sparse COO inputs.

## Code example

```python
>>> mask = torch.tensor([[0, 1], [1, 1]], dtype=torch.bool).to_sparse_csr()
>>> data = torch.tensor([[1, 2], [3, 0]]).to_sparse_csr()
>>> data.sparse_mask(mask)
tensor(crow_indices=tensor([0, 1, 2]),
       col_indices=tensor([1, 0]),
       values=tensor([2, 3]), size=(2, 2), nnz=2, layout=torch.sparse_csr)
```

The expected result is
```python
tensor(crow_indices=tensor([0, 1, 3]),
       col_indices=tensor([1, 0, 1]),
       values=tensor([2, 3, 0]), size=(2, 2), nnz=3, layout=torch.sparse_csr)
```
to meet the documented specification and the behavior with COO and strided inputs:
```python
>>> data.to_dense().sparse_mask(mask)
tensor(crow_indices=tensor([0, 1, 3]),
       col_indices=tensor([1, 0, 1]),
       values=tensor([2, 3, 0]), size=(2, 2), nnz=3, layout=torch.sparse_csr)
>>> data.to_sparse_coo().sparse_mask(mask)
tensor(crow_indices=tensor([0, 1, 3]),
       col_indices=tensor([1, 0, 1]),
       values=tensor([2, 3, 0]), size=(2, 2), nnz=3, layout=torch.sparse_csr)
```

## System Info

- PyTorch version: main

**Comment by Kousik-Sasmal** [comment]
I want to work on.

**Comment by Kousik-Sasmal** [comment]
Could I be assigned to work on this issue?

**Comment by pearu** [comment]
The current implementation defines `data.sparse_mask(mask)` as `data * mask`: https://github.com/pytorch/pytorch/blob/256fed02e930210dbcd7e5e23fcf142362098c2a/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp#L382
That means that the indices of the result are the intersection of the indices of `data` and `mask`. While this is expected for `mul` operation, the result of `sparse_mask` must have the indices of `mask`. So, the fix to this issue is to implement a projection kernel that maps the indices of `data` to the indices of `mask` (indices not in `mask` will be ignored, indices not in `data` will materialize the corresponding zeros).

@Kousik-Sasmal are you up to implementing the projection kernel as described above?

**Comment by Kousik-Sasmal** [comment]
> 

@pearu Yes, I understood the issue and working on that.

**Comment by pearu** [comment]
Btw, a quick fix would be to use:
```python
return self.to_dense().sparse_mask(mask);
```
instead of
https://github.com/pytorch/pytorch/blob/c5ad44be1d4d39a9cd5f93013f4f0f28fd1ef0f8/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp#L417
